File: 2024/3.py
import time, re, math

from utils import readInput

def loadInput(filename: str) -> list[str]:
    return readInput(filename)
    
def part1(memory: list[str]) -> None:
    r = re.compile(r"mul\((\d{1,3}),(\d{1,3})\)")
    tot = 0
    for mem in memory:
        matches = r.findall(mem)
        for m in matches:
            tot += math.prod(list(map(int, m)))
    print (f"🎄 Part 1: {tot}")

def part2(memory: list[str]) -> None:
    r = re.compile(r"mul\(\d{1,3},\d{1,3}\)|do\(\)|don't\(\)")
    enabled = True
    tot = 0
    for mem in memory:
        matches = r.findall(mem)
        for m in matches:
            if m == "do()":
                enabled = True
            elif m == "don't()":
                enabled = False
            else:
                if enabled:
                    m = list(map(int, m[4:-1].split(",")))
                    tot += math.prod(list(map(int, m)))
    print (f"🎄🎅 Part 2: {tot}")

if __name__ == "__main__":
    title = "Day 3: Mull It Over"
    sub = "⛄"*(len(title)//2-1+2)

    print()
    print(f" {title} ")
    print(sub)
    
    inputs = loadInput("input_3.txt")
    
    t0 = time.time()
    part1(inputs)
    print ("Time: {:.5f}".format(time.time()-t0))
    
    t0 = time.time()
    part2(inputs)
    print ("Time: {:.5f}".format(time.time()-t0))

File: 2024/3.jl
function loadInput(filename::String)
    readlines(filename)
end

function part1(memory)
    re = r"mul\((\d{1,3}),(\d{1,3})\)"
    tot = 0
    for mem in memory
        for match in eachmatch(re, mem)
            tot += prod(parse.(Int, match.captures))
        end
    end
    println("🎄 Part 1: $tot")
end

function part2(memory)
    re = r"mul\((\d{1,3}),(\d{1,3})\)|do\(\)|don't\(\)"
    enabled = true
    tot = 0
    for mem in memory
        for match in eachmatch(re, mem)
            m = match.match
            if m == "do()"
                enabled = true
            elseif m == "don't()"
                enabled = false
            else
                if enabled
                    tot += prod(parse.(Int, match.captures))
                end
            end
        end
    end
    println("🎄🎅 Part 2: $tot")
end

function main()
    title = "Day 3: Mull It Over"
    sub = "❄ " ^(length(title) ÷ 2 + 2)

    println()
    println(" $title ")
    println(sub)

    inputs = loadInput("input_3.txt")

    @time part1(inputs)
    @time part2(inputs)
end

main()
